### PR TITLE
Throw errors for invalid values in panner node constructor

### PR DESCRIPTION
**Pull request by Manishearth**
This was an oversight in the spec and is being fixed

https://github.com/WebAudio/web-audio-api/issues/1728

r? @ferjm

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/21555)

### CODE REVIEW

File: components/script/dom/pannernode.rs [hunk 1]
```diff
@@ -60,6 +60,15 @@ impl PannerNode {
         if count > 2 {
             return Err(Error::NotSupported)
         }
+        if options.maxDistance < 0. {
+            return Err(Error::NotSupported)
```
Comment (ferjm) [inline components/script/dom/pannernode.rs]: Should this be a `RangeError`? https://webaudio.github.io/web-audio-api/#dom-pannernode-maxdistance

File: components/script/dom/pannernode.rs [hunk 2]
```diff
@@ -60,6 +60,15 @@ impl PannerNode {
...
+        }
+        if options.rolloffFactor < 0. {
+            return Err(Error::Range("rolloffFactor should be positive".into()))
+        }
+        if options.coneOuterGain < 0. || options.coneOuterGain > 360. {
```
Comment (ferjm) [inline components/script/dom/pannernode.rs]: And this one `InvalidStateError`? https://webaudio.github.io/web-audio-api/#dom-pannernode-coneoutergain

**Review by ferjm — APPROVED** [review]
Thanks! It seems that https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/webaudio/the-audio-api/the-pannernode-interface/ctor-panner.html is not checking these constraints. So it'd be nice to add the additional checks.

**Comment by Manishearth** [comment]
Added tests.

r? @ferjm

File: components/script/dom/pannernode.rs [hunk 3]
```diff
@@ -54,16 +54,29 @@ impl PannerNode {
...
             return Err(Error::NotSupported)
         }
-        if count > 2 {
+        if count > 2 || count == 0 {
             return Err(Error::NotSupported)
         }
+        if *options.maxDistance <= 0. {
+            return Err(Error::Range("maxDistance should be positive".into()))
+        }
+        if *options.refDistance < 0. {
```
Comment (ferjm) [inline components/script/dom/pannernode.rs]: Perhaps `<= 0.`. The spec says to throw when it is set to a non-positive value. Same as with `maxDistance`

File: components/script/dom/pannernode.rs [hunk 4]
```diff
@@ -230,19 +243,23 @@ impl PannerNodeMethods for PannerNode {
         Finite::wrap(self.ref_distance.get())
     }
     // https://webaudio.github.io/web-audio-api/#dom-pannernode-refdistance
-    fn SetRefDistance(&self, val: Finite<f64>) {
+    fn SetRefDistance(&self, val: Finite<f64>) -> Fallible<()> {
+        if *val < 0. {
```
Comment (ferjm) [inline components/script/dom/pannernode.rs]: Same comment as before: `<= 0.`

File: components/script/dom/pannernode.rs [hunk 5]
```diff
@@ -289,7 +306,7 @@ impl PannerNodeMethods for PannerNode {
     }
     // https://webaudio.github.io/web-audio-api/#dom-pannernode-coneoutergain
     fn SetConeOuterGain(&self, val: Finite<f64>) -> Fallible<()> {
-        if *val < 0. || *val > 360. {
```
Comment (ferjm) [inline components/script/dom/pannernode.rs]: Ugh, I missed this one. Sorry.

**Review by ferjm — APPROVED** [review]
r=me. Thanks!

**Comment by Manishearth** [comment]
So the spec says that both need to be non-positive, but  [the test](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/webaudio/the-audio-api/the-pannernode-interface/panner-distance-clamping.html) indicates that zero is allowed for refDistance (but not maxDistance), and this agrees with Chrome. Firefox doesn't seem to throw errors at all?

I'll land this as is until [the spec folks can determine what it should be](https://github.com/WebAudio/web-audio-api/issues/1728#issuecomment-417710437)

@bors-servo r=ferjm

**Comment by Manishearth** [comment]
Spec editor confirmed that a zero value is okay for ref distance, I'll open a PR later.

**Pull request merged**